File: AlphaBetaPruningBayesian/AlphaBetaPruningBayesian/GameTree.fs
namespace AlphaBetaPruningBayesian

open System

type public GameTree<'T when 'T :> IGame>(initialState : 'T, childStatesGenerator : 'T -> 'T[]) = 
    member val public CurrentState = 
        initialState
        with get, set
       
    member val public ChildStatesGenerator = 
        childStatesGenerator
        with get, set

    member public this.HasChildStates() = 
        (this.ChildStatesGenerator this.CurrentState).Length > 0

    member public this.FindBestMove(depth) =
        let childStates = this.ChildStatesGenerator this.CurrentState

        if childStates.Length = 0 then
            Unchecked.defaultof<'T>
        else
            let mutable bestMove = childStates.[0]

            for childState in childStates do
                let miniMaxValue = this.MiniMax(childState, depth - 1, true, Int32.MinValue, Int32.MaxValue)

                if miniMaxValue < bestMove.StaticEvaluation() then
                    bestMove <- childState

            bestMove

    member public this.Run(depth, maximizingPlayer) =
        this.MiniMax(this.CurrentState, depth, maximizingPlayer, Int32.MinValue, Int32.MaxValue)

    member private this.MiniMax(state, depth, maximizingPlayer, alpha, beta) = 
        let childStates = this.ChildStatesGenerator state

        if depth <= 0 || childStates.Length = 0 then
            state.StaticEvaluation()
        else 
            let mutable miniMaxValue = if maximizingPlayer then Int32.MinValue else Int32.MaxValue

            let mutable newAlpha = alpha
            let mutable newBeta = beta
        
            let mutable valueFound = false

            for childState in childStates do
                if not valueFound then
                    let childMiniMax = this.MiniMax(childState, depth - 1, not maximizingPlayer, alpha, beta)

                    if maximizingPlayer then
                        miniMaxValue <- max miniMaxValue childMiniMax

                        newAlpha <- max newAlpha childMiniMax
                    else
                        miniMaxValue <- min miniMaxValue childMiniMax

                        newBeta <- min beta childMiniMax

                    if newBeta <= newAlpha then
                        valueFound <- true

            miniMaxValue
                

File: AlphaBetaPruning/AlphaBetaPruning/GameTree.cs
using System;
using System.Text;

namespace AlphaBetaPruning
{
    public class GameTree<T> where T : IGame
    {
        public T CurrentState { get; set; }

        public Func<T, T[]> ChildStatesGenerator { get; }

        public GameTree(T currentState, Func<T, T[]> childStatesGenerator)
        {
            CurrentState = currentState;
            ChildStatesGenerator = childStatesGenerator;
        }

        public bool HasChildStates() => ChildStatesGenerator.Invoke(CurrentState).Length > 0;
        
        public T FindBestMove(int depth)
        {
            T[] childStates = ChildStatesGenerator.Invoke(CurrentState);

            if (childStates.Length == 0)
            {
                return default;
            }

            T bestMove = childStates[0];

            foreach(T childState in childStates)
            {
                int miniMaxValue = MiniMax(childState, depth, false);

                if (miniMaxValue > bestMove.StaticEvaluation(false))
                {
                    bestMove = childState;
                }
            }

            return bestMove;
        }

        public int Run(int depth, bool maximizingPlayer)
        {
            return MiniMax(CurrentState, depth, maximizingPlayer);
        }

        private int MiniMax(T state, int depth, bool maximizingPlayer, int alpha = int.MinValue, int beta = int.MaxValue)
        {
            T[] childStates = ChildStatesGenerator.Invoke(state);

            if (depth <= 0 || childStates.Length == 0)
            {
                return state.StaticEvaluation(maximizingPlayer);
            }

            int miniMaxValue = maximizingPlayer ? int.MinValue : int.MaxValue;
            foreach(T childState in childStates)
            {
                int childMiniMax = MiniMax(childState, depth - 1, !maximizingPlayer, alpha, beta);
                if (maximizingPlayer)
                {
                    miniMaxValue = Math.Max(miniMaxValue, childMiniMax);

                    alpha = Math.Max(alpha, childMiniMax);
                }

                if (!maximizingPlayer)
                {
                    miniMaxValue = Math.Min(miniMaxValue, childMiniMax);

                    beta = Math.Min(beta, childMiniMax);
                }

                if (beta <= alpha)
                {
                    break;
                }
            }

            return miniMaxValue;
        }

        public void Print()
        {
            StringBuilder sb = new StringBuilder();

            System.Console.WriteLine("Game Tree");
            
            int indentationSize = 0;

            AppendString(CurrentState, true);

            void AppendString(T currentState, bool maximizingPlayer)
            {
                Console.ForegroundColor = (ConsoleColor)indentationSize;
                Console.WriteLine(new string(' ', indentationSize * 2) + $"{currentState}(ev. {MiniMax(currentState, 1000, maximizingPlayer)}[{maximizingPlayer}])");
                Console.ForegroundColor = ConsoleColor.Gray;
                indentationSize++;

                foreach(T childState in ChildStatesGenerator.Invoke(currentState))
                {
                    AppendString(childState, !maximizingPlayer);
                }

                indentationSize--;
            }
        }
    }
}
